### PR TITLE
Trinity oauth backend

### PR DESCRIPTION
**Pull request by melvinsoft**

### CODE REVIEW

**Comment by melvinsoft** [comment]
[(4) Trinity: SSO - steps before go to production](https://trello.com/c/YZ7UlMC9/1891-4-trinity-sso-steps-before-go-to-production)

File: social/backends/trinity.py [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+    #DEFAULT_SCOPE = ['email']
+    EXTRA_DATA = [
+        ('email', 'email'),
+        ('username', 'username'),
+        ('fullname', 'fullname'),
+        ('district', 'district'),
+    ]
+
+    def get_user_details(self, response):
+        """Return user details from GitHub account"""
```
Comment (tkeemon) [inline social/backends/trinity.py]: You probably want to update this docstring.

File: social/backends/trinity.py [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+        fullname, first_name, last_name = self.get_user_names(
+            '', response.get('first_name', ''),response.get('last_name', '')
+        )
+        data = {'username': response.get('username'),
```
Comment (tkeemon) [inline social/backends/trinity.py]: Add a fallback value here, or, if `username` is required and not provided, we should throw an error.

File: social/backends/trinity.py [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+                'email': response.get('email') or '',
+                'fullname': fullname or '',
+                'district': response.get('original_district')}
```
Comment (tkeemon) [inline social/backends/trinity.py:29]: Ditto here. Let's either fall back to some default value or throw an error.

File: social/backends/trinity.py [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+
+    def user_data(self, access_token, *args, **kwargs):
+        """Loads user data from service"""
+        url = 'https://pass.texasgateway.org/api/v1/user/me?' + urlencode({
+            'access_token': access_token
+        })
+        return self.get_json(url)
+
+    def get_user_id(self, details, response):
+        """Use isc username as unique id"""
```
Comment (tkeemon) [inline social/backends/trinity.py]: Let's fix up this docstring too.

File: examples/django_example/example/settings.py [hunk 5]
```diff
@@ -201,7 +201,9 @@
     'social.backends.vend.VendOAuth2',
     'social.backends.email.EmailAuth',
     'social.backends.username.UsernameAuth',
+    'social.backends.trinity.TrinityOauth2',
```
Comment (tkeemon) [inline examples/django_example/example/settings.py]: We probably don't need to include this here unless we're pushing this upstream. Our use case seems a bit too specific for that.

File: social/backends/trinity.py [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+        return details['username']
+
+    def oauth_authorization_request(self, token):
```
Comment (tkeemon) [inline social/backends/trinity.py]: It looks like this method is only used by the BaseOAuth1 class. Is there some reason you're including it here?

**Review by tkeemon — COMMENTED** [review]
I've included a lot of pedantic changes. The PR looks good to me. Has this been tested against Trinity's backend yet?

File: social/backends/trinity.py (hunk 2 above)
Comment (melvinsoft) [reply social/backends/trinity.py]: @tkeemon Correct, the OAuth identity provider **must** to provide this value, if not the workflow must fail.

File: social/backends/trinity.py (hunk 3 above)
Comment (melvinsoft) [reply social/backends/trinity.py:29]: @tkeemon Same as above, whit out this value, the edX workflow will fail, so must be always present.

File: social/backends/trinity.py (hunk 1 above)
Comment (melvinsoft) [reply social/backends/trinity.py]: Done!

File: social/backends/trinity.py (hunk 4 above)
Comment (melvinsoft) [reply social/backends/trinity.py]: Done!

File: social/backends/trinity.py (hunk 6 above)
Comment (melvinsoft) [reply social/backends/trinity.py]: @tkeemon Great catch! It was supposed to work on this function, but wasn't necessary at the end. Removed!

File: examples/django_example/example/settings.py (hunk 5 above)
Comment (melvinsoft) [reply examples/django_example/example/settings.py]: Removed!

**Comment by melvinsoft** [comment]
@tkeemon Great catches, all requested changes addressed!

**Review by tkeemon — APPROVED** [review]
Thanks. Looks good.

:+1:

**Pull request merged**